File: boj/not_passed/4862.py
import sys
def period(m, n):
    chk = [0] * 10000000
    mod = pow(10, n)
    temp = m % mod
    m = temp
    i = 1
    while 1:
        if chk[temp]:
            return [i, chk[temp]]
        chk[temp] = i
        i += 1
        temp = temp * m % mod

b = int(sys.stdin.readline())
while b:
    i, n = int(sys.stdin.readline()), int(sys.stdin.readline())
    ans = b
    p = period(b, n)
    for j in range(i - 1):
        ans = p[1] + (ans - p[1]) % (p[0] - p[1])
        ans = pow(b, ans)
    ans %= pow(10, n)    
    print(str(ans).rjust(n, '0'))
    b = int(sys.stdin.readline())
        

File: boj/not_passed/4862.c
#include <stdio.h>
#include <string.h>
int chk[10000000];
int _pow(int n)
{
    int ret = 10;
    while(--n)
        ret *= 10;
    return ret;
}

int period(int *idx, int m, int n)
{
    int mod = _pow(n);
    long long temp = m % mod;
    m = (int)temp;
    memset(chk, 0, sizeof(chk));
    *idx = 1;
    while (1)
    {
        if (chk[temp])
            return chk[temp];
        chk[temp] = (*idx)++;
        temp = temp * m % mod;
    }
    return 0;
}

int mypow(int num, int exp, int start, int end)
{
    long long ret = 1;
    while(exp)
    {
        if (exp & 1)
            ret = (ret * num) % mod;
        num = (num * num) % mod;
        exp /= 2;
    }
    return ret;
}

int main()
{
   
    int b, n, i, j,start, end, ans;
    scanf("%d", &b);
    while(b)
    {
        scanf("%d%d", &i, &n);
        ans = b;
        start = period(&end, b, n);
        for (j = 0; j < i - 1; j++)
        {
            ans = start + (ans - start) % (end - start);
            ans = mypow(b, ans, end - start);            
        }
        ans %= _pow(n);
        printf("%d\n", ans);
        scanf("%d", &b);
    }  
}